### ISSUE TITLE
What happend when multiple faces rendered to the same pixel?

### ISSUE THREAD

**Issue by Yumin-Sun-00** [facebookresearch/pytorch3d#594]
## ❓What happend when multiple faces rendered to the same pixel?

I know that from rasterizer, we can get a pix_to_face map, that we know the nearest k faces to that pixel. But it only store a fix number of K nearest face. When multiple faces rendered to the same pixel, it is hard to decide such K globally. Is there a way, that we can store the mapping between pixel and faces? So that every face has at least 1 pixel assigned.

**Comment by nikhilaravi** [comment]
@Yumin-Sun-00 The rasterizer only cares about the the top K mapping between the image pixels and the faces not the other way around as you mention here. Are you concerned about faces not getting gradients if they are not part of the top K for any pixels? What's your use case for requiring this?

**Comment by Yumin-Sun-00** [comment]
> @Yumin-Sun-00 The rasterizer only cares about the the top K mapping between the image pixels and the faces not the other way around as you mention here. Are you concerned about faces not getting gradients if they are not part of the top K for any pixels? What's your use case for requiring this?

I need a bijective mapping when I project mesh to image, and map from image back to meshes. I can understood that rasterizer care only top K, but would that make sense, that actually each triangle is at least visited once during rasterization, so we should have the meta data how a triangle is mapped to which pixel?

**Comment by bottler** [comment]
The mapping between faces and pixels could be very far from a bijection. Some faces might not be in the field of view, in which case there is no pixel to link them to.

We could have a scenario where all the the faces are in front of the camera spanning the whole field of view, non intersecting, in parallel one behind the other. In such a case, what do you expect in the mapping? Would you want it to link every face with every pixel? You can get *all* such links between faces and pixels by setting K to the number of faces and running the rasterizer.

**Comment by Yumin-Sun-00** [comment]
> The mapping between faces and pixels could be very far from a bijection. Some faces might not be in the field of view, in which case there is no pixel to link them to.
> 
> We could have a scenario where all the the faces are in front of the camera spanning the whole field of view, non intersecting, in parallel one behind the other. In such a case, what do you expect in the mapping? Would you want it to link every face with every pixel? You can get _all_ such links between faces and pixels by setting K to the number of faces and running the rasterizer.

I am exactly at this corner case, I have a wonderful flat mesh in front of the camera, and no parallel overlapping. In this case, do you think it is possible for me hook from anywhere in the pytorch3d, to get such bijective  mapping?

**Comment by bottler** [comment]
Your corner case and mine our different. If you know you only care about your case, you can calculate the maximum K you need before the calculation - the maximum number of pixels you get per face. But it's still not bijective even in your case.

**Comment by Yumin-Sun-00** [comment]
> the maximum number of pixels you get per face.

There is such argument I can control max. pixels per face? I only know there is max. faces per pixel.

>  But it's still not bijective even in your case.

You mean there is for sure no way to modify some code to generate a bijective map of visible parts?

**Comment by Yumin-Sun-00** [comment]
> the maximum number of pixels you get per face.

There is such argument I can control max. pixels per face? I only know there is max. faces per pixel.

>  But it's still not bijective even in your case.

You mean there is for sure no way to modify some code to generate a bijective map of visible parts?

**Comment by bottler** [comment]
> the maximum number of pixels you get per face
I meant the maximum number of faces per pixel.

>You mean there is for sure no way to modify some code to generate a bijective map of visible parts?
Of course not. It would be impossible. When you say "bijective" you mean that each face is linked to one and only one pixel, and each pixel is linked to one and only one face, right?

**Comment by nikhilaravi** [comment]
Closing this issue as there hasn't been any further discussion. Please reopen it if you still have further questions.
